bls12_381/src/pairings.rs

```diff
@@ -276,6 +288,166 @@ impl<'a, 'b> Mul<&'b Scalar> for &'a Gt {
 impl_binops_additive!(Gt, Gt);
 impl_binops_multiplicative!(Gt, Scalar);
 
+impl<T> Sum<T> for Gt
+where
+    T: Borrow<Gt>,
+{
+    fn sum<I>(iter: I) -> Self
+    where
+        I: Iterator<Item = T>,
+    {
+        iter.fold(Self::identity(), |acc, item| acc + item.borrow())
+    }
+}
+
+impl Group for Gt {
+    type Scalar = Scalar;
+
+    fn random<R: RngCore + ?Sized>(rng: &mut R) -> Self {
+        loop {
+            let inner = Fp12::random(rng);
+
+            // Not all elements of Fp12 are elements of the prime-order multiplicative
+            // subgroup. We run the random element through final_exponentiation to obtain
+            // a valid element, which requires that it is non-zero.
+            if !bool::from(inner.is_zero()) {
+                return MillerLoopResult(inner).final_exponentiation();
```

Comment:
@daira: reminder to check whether we're missing the cyclotomic subgroup optimization for the final exp.